eng/pipelines/templates/steps/repo-mirror.yml

```diff
@@ -0,0 +1,125 @@
+parameters:
+  SOURCE_REPO: $(Build.Repository.Uri)
+  SOURCE_BRANCHES: $(Build.SourceBranchName)
+  TARGET_ENDPOINTS: ''
+  WORKING_DIRECTORY: ''
+
+steps:
+- script: mkdir -p ${{ parameters.WORKING_DIRECTORY }}
+  displayName: mkdir -p ${{ parameters.WORKING_DIRECTORY }}
+  condition: ne('${{ parameters.WORKING_DIRECTORY }}', '')
+- script: |
```

Comment:
I believe this script is way overcomplicated for what we are trying to do. It should be as simple as a clone and then a push for each repo you want to mirror to. We should not be trying to do any merges or rebases on the mirroring automation. We should always use a branch that only gets commits from the mirroring (i.e. a one-way mirror) otherwise we are setting up for a larger set of issues that I don't believe are necessary.

Reply:
Well the rebase is required in our solution. We need yo keep custom commit in our test repo while keeping them sync with upstream

Reply:
Actually this script works very well for months. Previously it was a simple one but O found out that a simple script cannot meet our needs....

Reply:
I'd like to better understand the scenario. Are you using this mirroring step to sync branches that have other changes? Anytime there is a conflict it will require manual intervention so I'd recommend against it if possible.

Reply:
As you can see here https://github.com/openapi-env-test/azure-rest-api-specs/commits/master we keep customized test changes on top of all the commits so we could mange them easily. That's the feature we need in test env. Actually "git rebase --strategy-option=theirs" works pretty well in our sceneario.

Reply:
Perhaps we can talk a little more about why you would want to do that, especially in master. The rebase strategy will only work if you don't have any conflicting changes, if there is a conflict someone must resolve it or some changes may get lost if you blindly take theirs.

Reply:
Well this feature is only required for test repo and we should monitor the pipeline if any failure happens. That's not a problem for prod repos.

Reply:
Is there a reason that the test branches need to have the latest changes from master? What is it that generates the commits for the test repos? I think that would be the better place to do these rebase/merges.

Reply:
We do a PR mirror in our test repo to fast test our new pipelines/tools. In that case the master branch should be updated. Yes we should use better automated integration test for this, but we don't have enough time to build the test for it.

Reply:
OK I'd like to keep the pipelines/processes that are handling the testing scenario independent of the mirroring scenario. The general mirroring scenario should be a simple one-way mirror we can use my PR https://github.com/Azure/azure-sdk-tools/pull/680 to handle that. For the extra testing scenarios we can add another pipeline with this extra complexity and name/scope it appropriately for the testing purpose.

Reply:
Well in that case, we still need more repo like terraform sync. They do actually need to keep customized commit in https://github.com/Azure/azure-sdk-for-trenton/ ... IMO use two pipeline and config will increase maintenance effort because we to monitor two of them. I would like it to be a single dev tools that could do repo sync and help all the engineer.

Reply:
I do not have access to the azure-sdk-for-trenton repo so I don't know what that is for. Why would they need to have customized commits? I will restate my concern is that we shouldn't be trying to automatically mirror to a branch that has changes on both sides (i.e. two-way sync) as there will either have conflicts or worse data loss. We should keep the primary sync-ing pipeline simple and a one-way sync. Any repo that needs to have other commits needs to have independent automation setup for it because they need to be able to handle their particular workflow specific to their needs.

Reply:
As an example I would much rather the automation that is trying to do automatic merging or rebasing submit a PR to the repo with the update instead of automatically pushing to the repo that gives folks a chance to at least ensure that the changes look correct and CI jobs can validate.  Where as the one-way mirror can safely always be pushed by the bot account.